File: modules/client/src/main/java/com/eventhorizon/pulsar/client/packet/handler/PacketHandlerRegistry.java
package com.eventhorizon.pulsar.client.packet.handler;

import com.eventhorizon.pulsar.core.packet.Packet;
import com.eventhorizon.pulsar.core.packet.handler.PacketHandler;
import lombok.experimental.UtilityClass;
import org.jetbrains.annotations.NotNull;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.CopyOnWriteArrayList;

@UtilityClass
public final class PacketHandlerRegistry {

    private static final Map<String, List<PacketHandler<?>>> REGISTERED_HANDLERS = new ConcurrentHashMap<>();

    private static final Logger LOGGER = LoggerFactory.getLogger(PacketHandlerRegistry.class.getName());

    public static <T extends Packet> void register(@NotNull Class<T> packetClass, @NotNull PacketHandler<T> handler) {
        final var qualifiedName = packetClass.getName();

        REGISTERED_HANDLERS
                .computeIfAbsent(qualifiedName, k -> new CopyOnWriteArrayList<>())
                .add(handler);

        LOGGER.info("[PULSAR] A packet handler registered for {}", qualifiedName);
    }

    public static <T extends Packet> void unregister(Class<T> packetClass, PacketHandler<T> handler) {
        final var qualifiedName = packetClass.getName();
        final var handlers = REGISTERED_HANDLERS.get(qualifiedName);

        if (handlers != null) {
            if (handlers.remove(handler)) {
                LOGGER.info("[PULSAR] A packet handler unregistered for: {}", qualifiedName);
            }
        }
    }

    public static List<PacketHandler<?>> getAll(String packetName) {
        final var handlers = REGISTERED_HANDLERS.get(packetName);

        return handlers != null ?
                Collections.unmodifiableList(handlers) :
                Collections.emptyList();
    }
}

File: modules/client/src/main/kotlin/com/eventhorizon/pulsar/client/packet/handler/PacketHandlerRegistry.kt
package com.eventhorizon.pulsar.client.packet.handler

import com.eventhorizon.pulsar.core.packet.Packet
import com.eventhorizon.pulsar.core.packet.handler.PacketHandler

@Suppress("UNCHECKED_CAST")
object PacketHandlerRegistry {
    val registeredHandlers = mutableMapOf<String, MutableList<PacketHandler<Packet>>>()

    inline fun <reified T : Packet> register(handler: PacketHandler<T>) {
        val qualifiedName = T::class.qualifiedName
            ?: error("Can't register packet handler because qualified class name is null.")

        val packetHandlers = registeredHandlers.getOrPut(qualifiedName) {
            mutableListOf()
        } as MutableList<PacketHandler<T>>

        packetHandlers.add(handler)
        println("[PULSAR] A new packet handler registered to packet: $qualifiedName")
    }

    inline fun <reified T : Packet> unregister(handler: PacketHandler<T>) {
        val qualifiedName = T::class.qualifiedName
            ?: error("Can't unregister packet handler because qualified class name is null.")

        val packetHandlers = registeredHandlers[qualifiedName]
                as MutableList<PacketHandler<T>>? ?: return

        packetHandlers.remove(handler)

        println("[PULSAR] An packet handler unregistered to packet: $qualifiedName")
    }

    fun getAll(packetName: String) : List<PacketHandler<*>> {
        return registeredHandlers[packetName]?.toList() ?: emptyList()
    }
}